### PR TITLE
 feat: add YOLO SWAG solid-waste detection model

### PR DESCRIPTION
**Pull request by itisacloud**
## Summary

  Adds the YOLO SWAG solid-waste model to fAIr as a 5 m grid-cell classifier.

  - Adds the training, prediction, preprocessing, and postprocessing pipeline
  - Adds the model STAC item, documentation, Docker runtime, and inference image
  - Aligns generated model cells with the source-image grid used for labels
  - Documents imagery/zoom requirements and local training workflow
  
 ## Validation

  - [x] `just setup`
  - [x] `just validate`
  - [x] `just test` — 383 passed, 1 skipped
  - [x] `just build yolo_swag_waste_grid_segmentation`
  - [x] `just example yolo_swag_waste_grid_segmentation` (with the banepa sample dataset, non solid waste)
  - [x] `just test-serve yolo_swag_waste_grid_segmentation` (with the banepa sample dataset, non solid waste)

### CODE REVIEW

File: models/yolo_swag_waste_grid_segmentation/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+"""Deterministic toy chips and labels for waste-grid pipeline tests."""
+
+from __future__ import annotations
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/tests/conftest.py:3]: Not needed we use python > 3.12

File: models/yolo_swag_waste_grid_segmentation/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+COPY data/sample /tmp/fair-src/data/sample
+
+RUN --mount=type=cache,target=/root/.cache/uv \
+    uv pip install hatchling hatch-vcs
+RUN --mount=type=cache,target=/root/.cache/uv \
+    python -m pip install \
+    --index-url https://download.pytorch.org/whl/cpu \
+    torch==2.10.0+cpu \
+    torchvision==0.25.0+cpu
+RUN --mount=type=cache,target=/root/.cache/uv \
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/Dockerfile:24]: we need to pin the libraries here in the dockerfile to make sure it won't be broken when there is library update , so lets pin rasterio pyproj geopandas !

File: models/yolo_swag_waste_grid_segmentation/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    --no-build-isolation \
+    "ultralytics>=8.4.0" \
+    rasterio \
+    pyproj \
+    geopandas \
+    shapely \
+    Pillow \
+    onnx \
+    onnxscript \
+    /tmp/fair-src[k8s,serve]
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/Dockerfile]: not needed anymore : 

```
fair-py-ops[k8s,serve]==0.3.5
``` 
Best to put it in arg ( the version )

File: models/yolo_swag_waste_grid_segmentation/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,95 @@
+# Builder stage: install dependencies
+FROM ghcr.io/astral-sh/uv:python3.13-trixie-slim AS builder
+ENV UV_SYSTEM_PYTHON=1 UV_LINK_MODE=copy
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential libgdal-dev gdal-bin libgl1 libglib2.0-0 \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY pyproject.toml README.md fair_zenml_patch.pth /tmp/fair-src/
+COPY fair /tmp/fair-src/fair
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/Dockerfile:11]: Both lines are not needed

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+            data=str(yolo_dir),
+            epochs=epochs,
+            batch=batch_size,
+            imgsz=chip_size,
+            device=device,
+            lr0=learning_rate,
+            weight_decay=weight_decay,
+            optimizer=optimizer,
+            freeze=freeze_layers,
+            cos_lr=scheduler == "cosine",
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]: Typo ? because it will only pass True or False ? in the model train param  , intented ? I dont think this will allow changing ?

File: models/yolo_swag_waste_grid_segmentation/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+## Pretrained source
+
+The pretrained weights, labels, and preparation utilities come from the [YOLO Solid Waste Assessment on Grids source repository](https://github.com/GIScience/solid-waste-detection-for-fAIr). Its training data comprises 60 OpenAerialMap scenes; the source repository is released under the MIT License. The model is based on the accompanying [SWAG preprint](https://doi.org/10.48550/arXiv.2605.02316). Review the licence of each source image before reusing training imagery.
+
+## Training coverage and examples
+
+The pretrained SWAG dataset covers geographically diverse OpenAerialMap scenes. The map below shows the source training-scene distribution.
+
+[![World map of SWAG training-scene locations](https://raw.githubusercontent.com/GIScience/solid-waste-detection-for-fAIr/dc826f335b5403ed8161f0b6f5656e89aae9b2b8/data/images/Overview.png)](https://github.com/GIScience/solid-waste-detection-for-fAIr/blob/dc826f335b5403ed8161f0b6f5656e89aae9b2b8/data/images/Overview.png)
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/README.md:15]: I cloudn't find the pretraining data ? do you hae zenedo links or hf link somewhere ?

File: models/yolo_swag_waste_grid_segmentation/stac-item.json [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    -180,
+    -90,
+    180,
+    90
+  ],
+  "properties": {
+    "datetime": "2024-01-01T00:00:00Z",
+    "created": "2026-07-20T00:00:00Z",
+    "updated": "2026-07-20T00:00:00Z",
+    "title": "YOLO Solid Waste Analysis on spatial Grids",
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/stac-item.json]: Solid Waste Detection 

Generic title is good  !

File: models/yolo_swag_waste_grid_segmentation/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+COPY infra/compose/docker-compose.yml infra/compose/init-db.sql /tmp/fair-src/infra/compose/
+COPY stacks/compose.yaml /tmp/fair-src/stacks/
+COPY data/sample /tmp/fair-src/data/sample
+
+RUN --mount=type=cache,target=/root/.cache/uv \
+    uv pip install hatchling hatch-vcs
+RUN --mount=type=cache,target=/root/.cache/uv \
+    python -m pip install \
+    --index-url https://download.pytorch.org/whl/cpu \
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/Dockerfile]: requires cuda here , not the cpu version because training requires gpu

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+        grid = build_grid_gdf(projected_bounds(bounds, mosaic_crs, target_crs), cell_size_m, target_crs)
+        utm_to_mosaic = Transformer.from_crs(target_crs, mosaic_crs, always_xy=True)
+        label_polygons = load_labels(label_file, target_crs)
+        waste_polygons = label_polygons[label_polygons["label"] == 1]
+        if waste_polygons.empty:
+            msg = "At least one label=1 waste polygon is required"
+            raise ValueError(msg)
+        waste_union = waste_polygons.geometry.union_all()
+        background_polygons = label_polygons[label_polygons["label"] == 0]
+        if background_polygons.empty:
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]: This line should be removed , because if background polygons are empty we want to assume all the background pixels except the solid cells are background 

it should be like this : 

```python
  background_polygons = label_polygons[label_polygons["label"] == 0]
  background_union = (
      None if background_polygons.empty else background_polygons.geometry.union_all()
  )
```

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+
+    waste = labels[labels["label"] == 1]
+    background = labels[labels["label"] == 0]
+    return waste.geometry.union_all(), None if background.empty else background.geometry.union_all()
+
+
+def split_cells(cells, source_polygons, val_ratio: float, test_ratio: float, seed: int) -> dict[int, str]:
+    """Split complete source label-polygon groups."""
+    groups_by_polygon: dict[int, list[int]] = {}
+    for cell_id, cell in cells.iterrows():
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]:
```suggestion
    for cell_id, cell in cells.iterrows():
           if source_polygons.empty:
                groups_by_polygon[cell_id] = [cell_id]
                continue
```

Comment (kshitijrajsharma) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: something like this to add all cell that are not waste are background

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 11]
```diff
@@ -0,0 +1,760 @@
...
+
+def build_mosaic(chip_paths: list[Path]) -> Path:
+    import subprocess
+
+    out_dir = Path(tempfile.mkdtemp(prefix="yolo_cls_mosaic_"))
+    vrt_path = out_dir / "mosaic.vrt"
+    filelist = out_dir / "chips.txt"
+    filelist.write_text("\n".join(str(p) for p in chip_paths))
+    subprocess.run(
+        ["gdalbuildvrt", "-input_file_list", str(filelist), str(vrt_path)],
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]: This is a problem for distroless inference images , I don't think you have gdal in your distroless inference image ! try to avoid using subprocess inside python pipeline ; it will be harder to debug on future if anything goes wrong and also makes inference docker image heavy

Comment (kshitijrajsharma) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: May be try to do it with rasterio !

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 12]
```diff
@@ -0,0 +1,760 @@
...
+            (subset / sidecar.name).symlink_to(sidecar)
+    return str(subset)
+
+
+def pick_utm_crs(lon: float, lat: float):
+    from rasterio.crs import CRS
+
+    zone = int((lon + 180) // 6) + 1
+    epsg = (32600 if lat >= 0 else 32700) + zone
+    return CRS.from_epsg(epsg)
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]: this function can directly use crs projection and this math is not required ? proj crs should be able to do conversion easily !

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 13]
```diff
@@ -0,0 +1,760 @@
...
+    import numpy as np
+    import rasterio
+    from PIL import Image
+    from pyproj import Transformer
+    from shapely.geometry import mapping
+    from shapely.ops import transform as shapely_transform
+
+    from fair.utils.data import resolve_directory
+
+    if "confidence_threshold" not in params:
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py]: not needed , if you specific confidence thrshold is require din stac item its always gonna be there and would be asserted

File: models/yolo_swag_waste_grid_segmentation/pipeline.py [hunk 14]
```diff
@@ -0,0 +1,760 @@
...
+            lr0=learning_rate,
+            weight_decay=weight_decay,
+            optimizer=optimizer,
+            freeze=freeze_layers,
+            cos_lr=scheduler == "cosine",
+            verbose=False,
+        )
+        if results and hasattr(results, "results_dict"):
+            top1 = results.results_dict.get("metrics/accuracy_top1", 0.0)
+            log_metadata(metadata={"accuracy_top1": float(top1), "epoch": epochs})
```
Comment (kshitijrajsharma) [inline models/yolo_swag_waste_grid_segmentation/pipeline.py:660]: this top 1 accuracy : i have a fear it might get overwhelmed specially on these kind of model where background pixels are bound to be dominating, even random model will easily score more than 70 % accuracy in cases which only predicts cell as background , would it be possible to consider another accuracy metric to be reported here ? perhaps per class accuracy may be only for the label =1 ( for the wastes ) ? something like binary classification accuracy ( which neglects the background ? ) , may be having : 

```python
waste_recall = ... 
waste_precision = .. 
waste_f1=...
``` 
would be relevant here ? what do you think

File: models/yolo_swag_waste_grid_segmentation/Dockerfile (hunk 4 above)
Comment (kshitijrajsharma) [reply models/yolo_swag_waste_grid_segmentation/Dockerfile:11]: Reference here : https://github.com/hotosm/fAIr-models/blob/develop/models/dinov3s_buildings/Dockerfile , we recently moved away form editable install to pypi based install !

**Review by kshitijrajsharma — CHANGES_REQUESTED** [review]
Thanks for the PR , I have tested it and overall it looks great : 

Bravo on the spatial splitting , It splits cleanly the waste class however there might be some leakage on the background but I don't think it matters too much as our target class is already separated cleanly ! 

Tiny changes are required before I can merge :
 
- CUDA based docker image is required for training ( currently its CPU ) 
- gdal should not be included in the inference image and related code needs to be migrated either to use rasterio or to use any light library because this is the image that will be used to run live prediction on the browser
- labels should assume the background if the target class is not present ( I will also update documentation for this ) 
- choice of accuracy metric needs revisit ( I am slightly underconfident on this ) 

Kindly sync the changes with develop so that you will get new CI actions , Once all CI gets passed we are good to go ! 

Bravo for your great work !

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 5 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: no it gets evaluated based on the defined scheduler, renamed the var to be more indicative

File: models/yolo_swag_waste_grid_segmentation/README.md (hunk 6 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/README.md:15]: hey, we added a zenodo link

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 11 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: we got, rid of gdal in the function, sadly we cant build the virtual raster with rasterio but a physical should work equally

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 13 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: removed it, we are not 100% if our stac config requires it tho

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 12 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: switched to geopandas based crs

Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: also correct the transform also allows other configs and does not require rasterio crs object

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 14 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py:660]: our original thought, with both labeled waster and background should not be affected by this. We also think that including all the unlabled cells as background is more brittle in comparison to explicitly labeling background (stronger divide helps the model)

File: models/yolo_swag_waste_grid_segmentation/pipeline.py (hunk 10 above)
Comment (itisacloud) [reply models/yolo_swag_waste_grid_segmentation/pipeline.py]: we do not require the the scene to be completely labeled in the current approach, especially for larger scenes it can be helpfull to not have to large of a chunk of a single waste pile. also waste is rather sparse, wot
